apps/src/code-studio/progressRedux.js

```diff
@@ -472,328 +432,35 @@ export const setStageExtrasEnabled = stageExtrasEnabled => ({
-      level.uid === currentLevelId)
-  );
-};
-
-/**
- * The level object passed down to use via the server (and stored in stage.stages.levels)
- * contains more data than we need. This (a) filters to the parts our views care
- * about and (b) determines current status based on the current state of
- * state.levelProgress
- */
-const levelWithStatus = (
-  {levelProgress, levelPairing = {}, currentLevelId, isSublevel = false},
-  level
-) => {
-  if (level.kind !== LevelKind.unplugged && !isSublevel) {
-    if (!level.title || typeof level.title !== 'number') {
-      throw new Error(
-        'Expect all non-unplugged, non-bubble choice sublevel, levels to have a numerical title'
-      );
-    }
-  }
-  return {
-    ...processedLevel(level),
-    status: statusForLevel(level, levelProgress),
-    isCurrentLevel: isCurrentLevel(currentLevelId, level),
-    paired: levelPairing[level.activeId],
-    readonlyAnswers: level.readonly_answers
-  };
-};
-
-/**
- * Get level data for all lessons/stages
- */
-export const levelsByLesson = ({
-  stages,
-  levelProgress,
-  levelPairing,
-  currentLevelId
-}) =>
-  stages.map(stage =>
-    stage.levels.map(level => {
-      let statusLevel = levelWithStatus(
-        {levelProgress, levelPairing, currentLevelId},
-        level
-      );
-      if (statusLevel.sublevels) {
-        statusLevel.sublevels = level.sublevels.map(sublevel =>
-          levelWithStatus(
-            {levelProgress, levelPairing, currentLevelId, isSublevel: true},
-            sublevel
-          )
-        );
-      }
-      return statusLevel;
-    })
-  );
-
-/**
- * Get data for a particular lesson/stage
- */
-export const levelsForLessonId = (state, lessonId) =>
-  state.stages
-    .find(stage => stage.id === lessonId)
-    .levels.map(level => levelWithStatus(state, level));
-
-export const lessonExtrasUrl = (state, stageId) =>
-  state.stageExtrasEnabled
-    ? state.stages.find(stage => stage.id === stageId).lesson_extras_level_url
-    : '';
-
-export const isPerfect = (state, levelId) =>
-  !!state.levelProgress &&
-  state.levelProgress[levelId] >= TestResults.MINIMUM_OPTIMAL_RESULT;
-
-export const getPercentPerfect = levels => {
-  const puzzleLevels = levels.filter(level => !level.isConceptLevel);
-  if (puzzleLevels.length === 0) {
-    return 0;
-  }
-
-  const perfected = puzzleLevels.reduce(
-    (accumulator, level) =>
-      accumulator + (level.status === LevelStatus.perfect),
-    0
+  state.lessonGroups.some(group =>
+    group.lessons.some(lesson => lesson.lockable)
   );
-  return perfected / puzzleLevels.length;
-};
-
-/**
- * Given a level and levelProgress (both from our redux store state), determine
- * the status for that level.
- * @param {object} level - Level object from state.stages.levels
- * @param {object<number, TestResult>} levelProgress - Mapping from levelId to
- *   TestResult
- */
-export function statusForLevel(level, levelProgress) {
```

Comment:
(similar question as for bestResultLevelId) - Why are we able to get rid of this? Has the functionality changed?

Reply:
`userProgressFromServer` receives a dictionary of objects keyed by level id that contain some combination of 
```
{
status: string,
result: number,
paired: bool,
...
}
```
previously, we took just the `result` values from those objects and stored them in the `levelProgress` structure. then when we were joining our level data with progress data through `groupedLessons`, we would use the `result` value to compute the `status` string to store on the level. i believe this was in order to be backward compatible with the progress data we store in session storage, which only includes the `result` values (this related to the update from `MERGE_PROGRESS` to `MERGE_PROGRESS_RESULTS`).

now that we're storing full `studentLevelProgressType` objects instead of just `result` values, those objects contain the `status` string that we were previously adding to the levels, so we no longer need to compute it here.

removing this is entirely is based on the assumption that the `status` string being returned from the server is always correct, and doesn't need to be computed based on properties of the level. so far in my testing that has always appeared to be true, but if not, i'm hoping we'll find that in the bug bash.

Reply:
i should also note that for `MERGE_PROGRESS_RESULTS` where we ingest progress data from session storage with only `result` values, we do still need to compute the `status` string, which happens in `progressHelpers.levelProgressFromResult` using the `activityUtils.activityCssClass` helper function, which mirrors what we do on the server.

Reply:
Ahh thank you for the explanation. That makes a lot of sense.